src/util/focus-trap.ts

```diff
@@ -0,0 +1,142 @@
+   * @param _ngZone The zone Angular is running in
+   */
+  constructor(private _element: HTMLElement, private _document: Document, private _ngZone: NgZone) {
+    this._enforceFocus = this._enforceFocus.bind(this);
+    this._detectDirection = this._detectDirection.bind(this);
+
+    const eod = this._endOfDocument = this._document.createElement('i');
+    eod.className = 'ngb-focustrap-eod';
+    eod.tabIndex = 0;
+    this._document.body.appendChild(eod);
+
+    this._ngZone.runOutsideAngular(() => {
+      this._document.addEventListener('focusin', this._enforceFocus, true);
+      this._document.addEventListener('keydown', this._detectDirection);
+
+      this._removeDocumentListener = () => {
+        this._document.removeEventListener('focusin', this._enforceFocus, true);
+        this._document.removeEventListener('keydown', this._detectDirection);
+      }
+    });
+  }
+
+  /** Detect if incoming focus event should be prevented or not */
+  private _enforceFocus(event) {
+    const {target} = event;
+    if (this._document !== target && this._element !== target && !this._element.contains(target)) {
+      this._ngZone.run(() => {
+        const element = this.focusableElement;
+        if (element) {
+          element.focus();
+          event.stopPropagation();
+        }
+      });
+    }
+  }
+
+  /** Event handler detecting current `tab` navigation direction */
+  private _detectDirection(event) {
+    const {shiftKey, key} = event;
+    if (key === 'Tab') {
+      this._direction = shiftKey ? DIRECTION.BACKWARD : DIRECTION.FORWARD;
+    }
+  }
+
+  /** Try to set focus on the first found element that has an ngbAutofocus attribute */
+  private _focusInitial() {
+    const element = this._element.querySelector('[ngbAutofocus]') as HTMLElement;
+    if (element) {
+      element.focus();
+    }
+  }
+
+  /**
+   * Destroys the focustrap by removing all event listeners set on document.
+   *
+   * Eventually put the focus back on the previously focused element at the time
+   * focustrap has been initialized.
+   */
+  destroy() {
+    this._removeDocumentListener();
+    this._document.body.removeChild(this._endOfDocument);
+    if (this._previouslyFocused) {
+      this._previouslyFocused.focus();
+    }
+  }
+}
+
+/**
+ * Factory service to easily create a `NgbFocusTrap` instance on an element
+ */
+@Injectable()
+export class NgbFocusTrapFactory {
+  constructor(@Inject(DOCUMENT) private _document: Document, private _ngZone: NgZone) {}
+
+  /**
+   * Create an instance of {@link NgbFocusTrap} and return it
+   * @param element HTMLElement to trap focus inside
+   * @param autofocus Whether the focustrap should automatically move focus into the trapped element upon
+   * initialization and return focus to the previous activeElement upon destruction.
+   */
+  create(element: HTMLElement, autofocus = false): NgbFocusTrap {
+    const trap = new NgbFocusTrap(element, this._document, this._ngZone);
+    trap.autofocus = autofocus;
```

Comment:
Same question for autofocus setter. Do you see anybody calling it after constructing the service?

I mean doing `service = factory.create(el)` and then doing `service.autofocus = true` later, as opposed to `service = factory.create(el, true)` right away ?

Reply:
Yes, to me it was kind of related to one of the previous review comment.
I could see someone creating the focusTrap in `constructor`, and only activate autofocus in `ngAfterViewInit` to make sure the view element to focus is there.